### PR TITLE
Changes fourth type parameter of AbstractField to architecture

### PR DESCRIPTION
**Pull request by glwagner**
Previously, we used the type of `field.data` to infer the architecture of an `AbstractField`. However, this strategy fields for distributed models. In this PR we hold onto the architecture in a type parameter of `AbstractField`.

~~We may also want to add a field `architecture` to all fields that require confirmation if runtime information will ever be kept within architecture (for example, the layout of a distributed computation)... please advice @ali-ramadhan~~

We also add `architecture` as a property to all fields where `!isnothing(architecture(field))`.

### CODE REVIEW

**Review by ali-ramadhan — APPROVED** [review]
Looks great!

**Comment by ali-ramadhan** [comment]
I think GPU simulation and GPU regression test failures might have been intermittent (perhaps due to #1179?) so I restarted the build just in case.

File: src/Fields/abstract_field.jl [hunk 1]
```diff
@@ -21,7 +21,9 @@ import Oceananigans.Utils: datatuple
 Abstract supertype for fields located at `(X, Y, Z)` with data stored in a container
 of type `A`. The field is defined on a grid `G`.
 """
-abstract type AbstractField{X, Y, Z, A, G} end
+abstract type AbstractField{X, Y, Z,
+                            A <: Union{AbstractArchitecture, Nothing},
+                            G <: Union{AbstractGrid, Nothing}} end
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Fields/abstract_field.jl:26]: `ZeroField` has no architecture or grid. `FunctionField` has no architecture.

File: src/BuoyancyModels/buoyancy_field.jl [hunk 2]
```diff
@@ -14,33 +14,36 @@ import Oceananigans: short_show
...
+        return new{B, S, A, D, G, C}(data, arch, grid, buoyancy, tracers, status)
     end
 
-    function BuoyancyField(data, grid, buoyancy, tracers, status)
+    function BuoyancyField(data::D, arch::A, grid::G,
+                           buoyancy::B, tracers::C, status::S) where {D, A, G, B, C, S}
         validate_field_data(Center, Center, Center, data, grid)
-        return new{typeof(buoyancy), typeof(status), typeof(data),
-                   typeof(grid), typeof(tracers)}(data, grid, buoyancy, tracers, status)
+        return new{B, S, A, D, G, T}(data, grid, buoyancy, tracers, status)
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/BuoyancyModels/buoyancy_field.jl:46]: 🧹

**Pull request merged**